### PR TITLE
#18 - added token based authentication to the recommender api

### PR DESCRIPTION
**Pull request by lpuskas**
* added authentication middleware to the recommender (vault-bank libs added)
* authentication can be switched of with the -dev-mode=true flag
* required env vars are checked at start (viber added)
(VAULT_ADDR, TOKENSIGNINGKEY)
* tested manually, unit tests to come

### CODE REVIEW

File: api/routes.go [hunk 1]
```diff
@@ -55,6 +57,12 @@ func (r *RouteHandler) ConfigureRoutes(router *gin.Engine) {
 	}
 }
 
+// EnableAuth enables authentication middleware
+func (r *RouteHandler) EnableAuth(router *gin.Engine) {
+	signingKey := viper.GetString("tokensigningkey")
+	router.Use(auth.JWTAuth(auth.NewVaultTokenStore(""), signingKey, nil))
```
Comment (bonifaido) [inline api/routes.go]: The role should be an identifier which identifies the application, for example: `"telescopes"` later on we will configure Vault with that role.

Comment (lpuskas) [reply api/routes.go]: thanks @bonifaido, fxed

File: Gopkg.toml [hunk 2]
```diff
@@ -9,7 +9,7 @@
 
 [[constraint]]
   name = "github.com/gin-gonic/gin"
-  version = "1.2.0"
+  branch = "master"
```
Comment (stoader) [inline Gopkg.toml]: Do we need to switch to `master` branch?

Comment (lpuskas) [reply Gopkg.toml]: yes, as the added dependency (bak-vaults) depends on the master

File: main.go [hunk 3]
```diff
@@ -85,7 +123,18 @@ func main() {
 
 	routeHandler := api.NewRouteHandler(engine, api.NewValidator(providers))
 
+	// new default gin engine (recovery, logger middleware)
 	router := gin.Default()
+
+	// enable authentication if not dev-mode
+	if !*devMode {
+		log.Debug("enable authentication")
+		signingKey := viper.GetString("tokensigningkey")
```
Comment (stoader) [inline main.go]: use `cfgTokenSigningKey` constant instead of literal

File: main.go [hunk 4]
```diff
@@ -85,7 +123,18 @@ func main() {
...
+		appRole := viper.GetString("telescopes_app_role")
```
Comment (stoader) [inline main.go]: use `cfgAppRole` constant instead of literal

File: Gopkg.toml (hunk 2 above)
Comment (bonifaido) [reply Gopkg.toml]: The issue is that gin hasn't tagged a release a long time.

File: main.go (hunk 3 above)
Comment (lpuskas) [reply main.go]: thx, fixed

**Pull request merged**